File: Introduction_to_Robotics_Course/Chapter15/Lesson3/Chapter15_Lesson3.m
function R_sys = kOutOfNRel(Rc, n, k)
%KOUTOFNREL Reliability of a k-out-of-n architecture
%   Rc : scalar reliability of a single channel at time t
%   n  : total number of channels
%   k  : minimum number of working channels required

    arguments
        Rc (1,1) double {mustBeGreaterThanOrEqual(Rc,0), mustBeLessThanOrEqual(Rc,1)}
        n  (1,1) double {mustBeInteger, mustBePositive}
        k  (1,1) double {mustBeInteger, mustBePositive}
    end

    if k < 1 || k > n
        error('Need 1 <= k <= n');
    end

    R_sys = 0.0;
    for i = k:n
        c = nchoosek(n,i);
        R_sys = R_sys + c * Rc.^i .* (1.0 - Rc).^(n-i);
    end
end

% Example usage:
lambdaD = 1e-6;    % dangerous failures per hour
TI      = 8760.0;  % proof test interval (hours)
Rc      = exp(-lambdaD * TI);

R_1oo1  = kOutOfNRel(Rc, 1, 1);
R_1oo2  = kOutOfNRel(Rc, 2, 1);
disp([R_1oo1, R_1oo2])
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter15/Lesson3/Chapter15_Lesson3.java
public final class Reliability {

    public static double reliabilityExponential(double lambdaD, double t) {
        if (lambdaD < 0.0 || t < 0.0) {
            throw new IllegalArgumentException("lambdaD and t must be non-negative");
        }
        return Math.exp(-lambdaD * t);
    }

    public static double kOutOfN(double Rc, int n, int k) {
        if (Rc < 0.0 || Rc > 1.0) {
            throw new IllegalArgumentException("Rc must be in [0,1]");
        }
        if (k < 1 || k > n) {
            throw new IllegalArgumentException("Need 1 <= k <= n");
        }
        double Rsys = 0.0;
        for (int i = k; i <= n; ++i) {
            double c = comb(n, i);
            Rsys += c * Math.pow(Rc, i) * Math.pow(1.0 - Rc, n - i);
        }
        return Rsys;
    }

    private static double comb(int n, int r) {
        if (r < 0 || r > n) return 0.0;
        if (r == 0 || r == n) return 1.0;
        double c = 1.0;
        for (int i = 1; i <= r; ++i) {
            c *= (double)(n - r + i) / (double)i;
        }
        return c;
    }

    public static void main(String[] args) {
        double lambdaD = 1e-6;
        double TI = 8760.0;

        double Rc = reliabilityExponential(lambdaD, TI);
        double R1 = kOutOfN(Rc, 1, 1);
        double R2 = kOutOfN(Rc, 2, 1);

        System.out.println("Rc(T_I)  = " + Rc);
        System.out.println("R_1oo1   = " + R1);
        System.out.println("R_1oo2   = " + R2);
    }
}
      
